arangod/Aql/Functions.cpp

```diff
@@ -1288,6 +1347,232 @@ AqlValue Functions::Substring(arangodb::aql::Query* query,
+    auto pos = search->first(status);
+    int32_t len;
+    if (pos != USEARCH_DONE) {
+      len = search->getMatchedLength();
+    }
+    srchResultPtrs.push_back(std::make_pair(pos, len));
+  }
+
+  UnicodeString result;  
+  int32_t lastStart = 0;
+  int64_t count = 0;
+  while (true) {
+    int which = -1;
+    int32_t pos = USEARCH_DONE;
+    int32_t mLen = 0;
+    int i = 0;
+    for (auto resultPair : srchResultPtrs) {
+      // We locate the nearest matching search result.
+      int32_t thisPos;
+      thisPos = resultPair.first;
+      if ((pos == USEARCH_DONE) || (pos > thisPos)) {
+        if (thisPos != USEARCH_DONE) {
+          pos = thisPos;
+          which = i;
+          mLen = resultPair.second;
+        }
+      }
+      i++;
+    }
+    if (which == -1) {
+      break;
+    }
+    // from last match to this match, copy the original string.
+    result.append(unicodeStr, lastStart, pos - lastStart);
+    if (replace.size() != 0) {
+      if (replace.size() > (size_t) which) {
+        result.append(replace[which]);
+      }
+      else if (replaceWasPlainString) {
+        result.append(replace[0]);
+      }
+    }
+    
+    // lastStart is the place up to we searched the source string 
+    lastStart = pos + mLen;
+
+    // we try to search the next occurance of this string
+    auto& search = searchVec[which];
+    pos = search->next(status);
+    if (pos != USEARCH_DONE) {
+      mLen = search->getMatchedLength();
+    }
+    else {
+      mLen = -1;
+    }
+    srchResultPtrs[which] = std::make_pair(pos, mLen);
+
+    which = 0;
+    for (auto searchPair : srchResultPtrs) {
+      // now we invalidate all search results that overlap with
+      // our last search result and see whether we can find the
+      // overlapped pattern again.
+      // However, that mustn't overlap with the current lastStart
+      // position either.
+      int32_t thisPos;
+      thisPos = searchPair.first;
+      if ((thisPos != USEARCH_DONE) && (thisPos < lastStart)) {
+        auto &search = searchVec[which];
+        pos = thisPos;
+        while ((pos < lastStart) && (pos != USEARCH_DONE)) {
+          pos = search->next(status);
+          if (pos != USEARCH_DONE) {
+            mLen = search->getMatchedLength();
+          }
+          srchResultPtrs[which] = std::make_pair(pos, mLen);
+        }
+      }
+      which++;
+    }
+    
+    count ++;
+    if ((limit != -1) && (count >= limit)) {
+      // Do we have a limit count?
+      break;
+    }
+    // check whether none of our search objects has any more results
+    bool allFound = true;
+    for (auto resultPair : srchResultPtrs) {
+      if (resultPair.first != USEARCH_DONE) {
+        allFound = false;
+        break;
+      }
+    }
+    if (allFound) {
+      break;
+    }
+  }
+  // Append from the last found:
+  result.append(unicodeStr, lastStart, unicodeStr.length() - lastStart);
+  
+  result.toUTF8String(utf8);
+  return AqlValue(utf8);
+}
```

Comment:
I think the function is too long and should be split (No need to change it this time)